vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go

```diff
@@ -392,3 +431,149 @@ func setUpInformer(kubeClient kube_client.Interface, kind controllerKind) (cache
 	}
 	return informer, nil
 }
+
+// CanInPlaceUpdate performs the same checks
+func (e *podsEvictionRestrictionImpl) CanInPlaceUpdate(pod *apiv1.Pod) bool {
+	if !features.Enabled(features.InPlaceOrRecreate) {
+		return false
+	}
+	cr, present := e.podToReplicaCreatorMap[GetPodID(pod)]
+	if present {
+		// If our QoS class is guaranteed, we can't change the resources without a restart
+		if pod.Status.QOSClass == apiv1.PodQOSGuaranteed {
+			return false
+		}
+
+		if IsInPlaceUpdating(pod) {
+			return false
+		}
+
+		for _, container := range pod.Spec.Containers {
+			// If some of these are populated, we know it at least understands resizing
+			if container.ResizePolicy == nil {
+				klog.InfoS("Can't resize pod, container resize policy does not exist; is InPlacePodVerticalScaling enabled?", "pod", klog.KObj(pod))
+				return false
+			}
+		}
+
+		singleGroupStats, present := e.creatorToSingleGroupStatsMap[cr]
+		// If we're pending, we can't in-place resize
+		// TODO(jkyros): are we sure we can't? Should I just set this to "if running"?
```

Comment:
Good question, I imagine that a resize during pending is "less risky" as there's no running software/cgroup.

Reply:
Right, I'm not exactly sure what happens if you try this though. I will report back to see what happens, and act accordingly.

In theory we would want behaviour in a situation e.g. 
3 replica deployment
2 of them are running, but 1 is pending, and the 2 replicas are generating good recommendations that the pending pod would like to use while pending.

I'm assuming that's why this would be useful?

Reply:
> The thing is that the VPA updater seems to ignore pending pods for updates anyways, so I guess there's no point in this check at all? ([The podLister does not list pending pods](https://github.com/kubernetes/autoscaler/blob/52cd68a4987f875eb811afbe07b66d9d69f51a14/vertical-pod-autoscaler/pkg/updater/logic/updater.go#L176) at least for my kind cluster). Am I missing something? I thought the updater recognizes pending pods, since that's a condition in `CanEvict`...

Taking a look at this: https://github.com/kubernetes/autoscaler/blob/aa1d413ea3bf319b56c7b2e65ade1a028e149439/vertical-pod-autoscaler/pkg/updater/logic/updater.go#L299-L310

I assume that `"spec.nodeName!=" + ""` may exclude some Pending pods?

I wonder if this is something to change in the future?

Reply:
Yes, my understanding too is that `"spec.nodeName!=" + ""` will exclude pending pods.

Reply:
I think I should remove this check as well then.